netutils/password.py

```diff
@@ -233,6 +242,79 @@ def encrypt_type7(unencrypted_password: str, salt: t.Optional[int] = None) -> st
     return encrypted_password
 
 
+def _wpa_base64_encode(data: bytes):
```

Comment:
I don't think this is necessary, I think the same can be accomplished with a string translation table. I did some googling and found other implementations that seem more clear, such as [this](https://github.com/BrettVerney/ciscoPWDhasher/blob/74cc0989fb3d3c2802b0389811d1c8022a9ea41f/CiscoPWDhasher/__init__.py#L94).

Reply:
wow, that is a MUCH cleaner implementation. I wish I'd found that when i started researching this

Reply:
Ok, so i tested BrettVarney's base64 encoding implementation, and the hashes it produces do not match those produced by a cisco switch.

```console
! Tested on IOS-XE 17.3
(config)# username test algorithm-type scrypt secret cisco
(config)# do show run | include username
username test secret 9 $9$x2xAAwQ3MBbEnk$JCxr6MnPb.k5ymK72mTypyRJYH5W74ZRvtLTprCj.xQ
```

```python
>>> password.encrypt_type9_b64_translated("cisco", "x2xAAwQ3MBbEnk") 
'$9$x2xAAwQ3MBbEnk$jcxrWmN/BOKVyMkXSM3ypy1j8hV6XU91vtl3prcJOx0' # does not match cisco's show run
>>> password.encrypt_type9("cisco", "x2xAAwQ3MBbEnk")
'$9$x2xAAwQ3MBbEnk$JCxr6MnPb.k5ymK72mTypyRJYH5W74ZRvtLTprCj.xQ' # DOES match
```

Reply:
which confirms my prior research. Cisco implements the same bizarre base64 iterated encoding algorithm used by WPA to encode its passwords

Reply:
That's interesting, when I run this locally, it generates the same hash as in your IOS-XE example:

```python
#!/usr/bin/env python3

import hashlib
import base64

std_b64chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
cisco_b64chars = "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
b64table = str.maketrans(std_b64chars, cisco_b64chars)

def encrypt(password, salt):
    password_hash = hashlib.scrypt(password.encode(), salt=salt.encode(), n=16384, r=1, p=1, dklen=32)
    return base64.b64encode(password_hash).decode().translate(b64table)[:-1]

print(encrypt("cisco", "x2xAAwQ3MBbEnk"))
# JCxr6MnPb.k5ymK72mTypyRJYH5W74ZRvtLTprCj.xQ
```

Reply:
Some oddities with git, but just note @abates other comment about macos. It seems his solution should work, and there is a workaround for MacOS (basically to fail gracefully).

Reply:
@abates that is indeed interesting. your  code example is identical to the one i wrote to test base64 translation, except that my base64 translation table looked like this:

```python
str.maketrans(
    ALPHABET + '+/',
    ENCRYPT_TYPE9_ENCODING_CHARS
)
```

Reply:
Ahhhh, i see the discrepancy. `ALPHABET` as defined in `password.py` is composed of ascii_lowercase + ascii_uppercase + digits, not ascii_uppercase + ascii_lowercase + digits.

Ok, I'll re-implement

Reply:
reimplemented, and the tests are now passing. this base64 scheme behaves correctly